File: backend/app/schemas/content.py
from __future__ import annotations

from datetime import datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, Field

from ..models.content import ContentStatus


class AdminContentResponse(BaseModel):
    id: UUID
    title: str
    description: Optional[str]
    status: ContentStatus
    file_type: str
    file_size: Optional[int]
    category_id: Optional[UUID]
    owner_id: Optional[UUID]
    published_at: Optional[datetime]
    created_at: datetime
    updated_at: datetime
    file_path: str = Field(description="Relative path to stored content file")

    model_config = {"from_attributes": True}


class AdminContentListResponse(BaseModel):
    items: List[AdminContentResponse]


class ContentUpdateRequest(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    category_id: Optional[UUID] = None
    status: Optional[ContentStatus] = None


class ContentUpdateResponse(AdminContentResponse):
    message: str = "Content updated successfully"


class MemberContentResponse(BaseModel):
    id: UUID
    title: str
    description: Optional[str]
    file_type: str
    file_size: Optional[int]
    category_id: Optional[UUID]
    published_at: Optional[datetime]
    created_at: datetime
    owner_id: Optional[UUID]
    updated_at: datetime
    likes_count: int
    comments_count: int

    model_config = {"from_attributes": True}


class MemberContentListResponse(BaseModel):
    items: List[MemberContentResponse]
    total: int
    page: int
    page_size: int


class MemberContentDetailResponse(MemberContentResponse):
    status: ContentStatus


class ContentDownloadResponse(BaseModel):
    token: str
    expires_in: int = 300


class ContentCategoryResponse(BaseModel):
    id: UUID
    name: str
    description: Optional[str]

    model_config = {"from_attributes": True}


class ContentCategoryListResponse(BaseModel):
    items: List[ContentCategoryResponse]

File: frontend/src/lib/api/content.ts
import { apiClient, toApiError } from "@/lib/api/client";

export type ContentSummary = {
  id: string;
  title: string;
  description: string | null;
  file_type: string;
  file_size: number | null;
  category_id: string | null;
  published_at: string | null;
  created_at: string;
  updated_at: string;
  owner_id: string | null;
  likes_count: number;
  comments_count: number;
};

export type ContentListResponse = {
  items: ContentSummary[];
  total: number;
  page: number;
  page_size: number;
};

export type ContentCategory = {
  id: string;
  name: string;
  description: string | null;
};

export type ContentCategoryListResponse = {
  items: ContentCategory[];
};

export type ContentListParams = {
  page?: number;
  pageSize?: number;
  search?: string;
  categoryId?: string;
  contentType?: string;
  uploadedAfter?: string;
  uploadedBefore?: string;
};

export async function fetchContentList(params: ContentListParams = {}): Promise<ContentListResponse> {
  try {
    const { data } = await apiClient.get<ContentListResponse>("/content", {
      params: {
        page: params.page,
        page_size: params.pageSize,
        search: params.search || undefined,
        category_id: params.categoryId || undefined,
        content_type: params.contentType || undefined,
        uploaded_after: params.uploadedAfter || undefined,
        uploaded_before: params.uploadedBefore || undefined,
      },
    });
    return data;
  } catch (error) {
    throw toApiError(error);
  }
}

export async function fetchContentCategories(): Promise<ContentCategoryListResponse> {
  try {
    const { data } = await apiClient.get<ContentCategoryListResponse>("/content/categories");
    return data;
  } catch (error) {
    throw toApiError(error);
  }
}
